Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -225,14 +253,82 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
+                }
+                else
+                {
+                    // after deprecation: this will need to go when Session.Error is fully deprecated
+                    Session.RaiseError(session, exception);
+                }
             }
             catch (Exception ex)
             {
                 Logger.Default.Log(LogLevel.Warn, $"An error has occurred while handling a session error: {ex}");
             }
         }
 
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "Unowned Realm don't need to be disposed")]
+        private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
+        {
+            var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+            var syncConfiguration = (SyncConfigurationBase)syncConfigHandle.Target;
+
+            if (syncConfiguration.ClientResetHandler == null)
+            {
+                return true;
+            }
+
+            var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfiguration.ClientResetHandler;
+            var schema = syncConfiguration.GetSchema();
+            var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfiguration, schema);
+
+            try
+            {
+                discardLocalResetHandler.OnBeforeReset?.Invoke(realmBefore);
+            }
+            catch (Exception ex)
+            {
+                Logger.Default.Log(LogLevel.Warn, $"An error has occurred while executing SyncConfigurationBase.OnBeforeReset during a client reset: {ex}");
+                return false;
+            }
+
+            return true;
+        }
+
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyAfterClientReset))]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "Unowned Realm don't need to be disposed")]
+        private static bool NotifyAfterClientReset(IntPtr beforeFrozen, IntPtr after, IntPtr managedSyncConfigurationHandle)
+        {
+            var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+            var syncConfiguration = (SyncConfigurationBase)syncConfigHandle.Target;
+
+            // no clientResetHandler means don't do anything for this callback
+            if (syncConfiguration.ClientResetHandler == null)
+            {
+                return true;
+            }
+
+            var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfiguration.ClientResetHandler;
+            var schema = syncConfiguration.GetSchema();
+            var realmBefore = new Realm(new UnownedRealmHandle(beforeFrozen), syncConfiguration, schema);
+            var realmAfter = new Realm(new UnownedRealmHandle(after), syncConfiguration, schema);
```

Comment:
Same here - don't create these if the user hasn't specified `OnAfterReset` handler.